src/Lecture2.hs

```diff
@@ -175,64 +315,97 @@ challenging. However, you still may find some of them easier than some of the
 previous ones. Difficulty is a relative concept.
 -}
 
-{- | Write a function that takes a list of numbers and returns 'True'
-if all the numbers are in the increasing order (i.e. the list is
-sorted).
-
->>> isIncreasing [3, 1, 2]
-False
->>> isIncreasing [1 .. 10]
-True
--}
+-- | Write a function that takes a list of numbers and returns 'True'
+-- if all the numbers are in the increasing order (i.e. the list is
+-- sorted).
+--
+-- >>> isIncreasing [3, 1, 2]
+-- False
+-- >>> isIncreasing [1 .. 10]
+-- True
 isIncreasing :: [Int] -> Bool
-isIncreasing = error "TODO"
-
-{- | Implement a function that takes two lists, sorted in the
-increasing order, and merges them into new list, also sorted in the
-increasing order.
-
-The lists are guaranteed to be given sorted, so you don't need to
-verify that.
-
->>> merge [1, 2, 4] [3, 7]
-[1,2,3,4,7]
--}
+isIncreasing [] = True
+isIncreasing [_] = True
+isIncreasing [x, y] = x <= y -- Redundant but maybe useful performance wise
+isIncreasing (x : y : ys) = x <= y && isIncreasing (y : ys)
+
+-- | Implement a function that takes two lists, sorted in the
+-- increasing order, and merges them into new list, also sorted in the
+-- increasing order.
+--
+-- The lists are guaranteed to be given sorted, so you don't need to
+-- verify that.
+--
+-- >>> merge [1, 2, 4] [3, 7]
+-- [1,2,3,4,7]
 merge :: [Int] -> [Int] -> [Int]
-merge = error "TODO"
-
-{- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
-function takes a list of numbers and returns a new list containing the
-same numbers but in the increasing order.
-
-The algorithm of merge sort is the following:
-
-  1. If the given list has less than 2 elements, it's already sorted.
-  2. Otherwise, split list into two lists of the same size.
-  3. Sort each of two lists recursively.
-  4. Merge two resulting sorted lists to get a new sorted list.
-
->>> mergeSort [3, 1, 2]
-[1,2,3]
--}
+merge xs [] = xs
+merge [] ys = ys
+merge (x : xs) (y : ys)
+  | x == y = x : y : merge xs ys
+  | x < y = x : merge xs (y : ys)
+  | otherwise = y : merge (x : xs) ys
```

Comment:
You can use the standard function [compare](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Ord.html#t:Ord) in this case. It can be more efficient because it avoids multiple comparisons:

```suggestion
merge (x : xs) (y : ys) = case compare x y of
  EQ -> x : y : merge xs ys
  LT -> x : merge xs (y : ys)
  GT -> y : merge (x : xs) ys
```